### PR TITLE
Experiment class

### LINKED ISSUE #26: Data structure for cluster-dependent information
We need an efficient structure for per-cluster data.

* Based on 1D, 2D, or 3D NumPy arrays
* Cluster list on 1 (example of cluster statistics) or 2 axis (example of CCGs)
* Fast cluster indexing
* Fast update when the cluster assignments change
* Arbitrary cluster indices
* Relabelling

We'll probably need a dynamic array implementation on top of NumPy (inspired by [this](https://github.com/rougier/gl-agg/blob/master/glagg/dynamic_buffer.py) for example). For dual cluster axis (CCGs) we'll need something specific as well.

Ideally, this structure would contain a `cluster_map` variable with the cluster assignements for all spikes. When this variable is changed, the internal arrays are updated.

cc @nippoo

### LINKED ISSUE #27: IPython visualization widget with traitlets
Each view for clustering will be an IPython widget exposing specific traitlet attributes:

* clustering: a Clustering instance
* selected_spikes: a ndarray of selected spikes (selection used for highlighting or splitting)
* clusters: a list of selected clusters
* cluster_order: a string specifying the cluster order (by index, cluster group, size...?)

A base widget will implement those, custom widgets will derive from it.

In the final interface in IPython, we'll link all these traitlets together using IPython's `link()` function. When a spike selection changes in one widget, it will also change in the others.

To make this work, we'll need to implement specific traitlet types:

* ndarray (see [this](https://gist.github.com/maedoc/8187958))
* Clustering

Provisional list of clustering widgets:

* FeatureView
* WaveformView
* TraceView
* GridView
* CorrelogramsView
* SimilarityView

### LINKED ISSUE #31: Basic WaveformView
* Waveforms positioned with a probe geometry
* Subset of all spikes from a list of given clusters
* Point colors as a function of the cluster
* Implement traitlets so that selected spikes, cluster colors, and probe geometry can be easily changed through an API

### LINKED ISSUE #33: ndarray traitlet type

### LINKED ISSUE #46: phy.cluster.manual.color subpackage
* Facilities to generate distinct colors
* Generate a random color
* Generate a color distinct from a given color

(possibly: to be partially merged into VisPy later)

### LINKED ISSUE #48: Manual clustering Session object
Implement a user-level class with control actions:

```
class Session:
    def merge(clusters)
    def move(clusters, group)

    def undo()
    def redo()

    def start_wizard(self)
    def pause_wizard(self)
    def reset_wizard(self)
```

This class uses `Clustering`, `ClusterMetadata`, and `Selection` instances, and uses a `GlobalHistory` to track a unique undo stack with both clustering (merge, split, etc.) and cluster metadata (cluster moved, cluster color changed, etc.) actions.

This class can also update all views through the `Selection` instance. The different instances communicate with `UpdateInfo` instances.

### PR DESCRIPTION
**Pull request by rossant**
This PR paves the way for a prototype in the notebook. It prepares most of the infrastructure.
- Added an `Experiment` class that holds the data of an experiment. It is passed to the Session instance.
- The `Session` acts as a controller, managing all open views, the current experiment instance (with the current dataset), the clustering, and cluster metadata structures.
- Added a `MockExperiment`, useful for testing a session on artificial data.
- Added a new color module.
- Added simple system to store and load static CSS/JS files in the notebook.
- Updated the baking mechanism in `WaveformView`: GPU data is automatically baked when needed. Only what needs to be baked is baked.
- Added HTML/JS `ClusterView` view as an IPython widget.
- Added `enable_notebook()` function that enables GUI event loop integration and injects some JS and CSS in the notebook.

Refactoring:
- Added `_as_array()` to replace `np.asarray()`. The latter happily converts function instances to arrays with an `object` dtype, whereas we want to enforce numerical arrays.

I've tested the system in the notebook, with Qt waveform view, cluster view, selection.
- Closes #26
- Closes #27
- Closes #31
- Closes #33 (traitlets not needed with the current controller approach)
- Closes #46
- Closes #48

To do later:
- Highlight the selected clusters in the cluster view
- Multiple selection in the cluster view
- Find out how to unit test notebook logic on travis
- Test view interactivity
- Refactor baking logic so that other views can also benefit from it
- More views (features, correlograms)
- `KwikExperiment` with load/save features

### CODE REVIEW

**Comment by rossant** [comment]
@nippoo this PR is ready for review! It would be great you could have a look and possibly merge by Monday, so that we can discuss about what's next.

File: phy/datasets/mock.py [hunk 1]
```diff
@@ -21,10 +25,91 @@ def artificial_waveforms(n_spikes=None, n_samples=None, n_channels=None):
...
+    return np.cumsum(nr.randint(low=0, high=100, size=n_spikes))
+
+
+#------------------------------------------------------------------------------
+# Artificial Experiment
+#------------------------------------------------------------------------------
+
+class MockExperiment(BaseExperiment):
+    n_channels = 32
+    n_features = 32 * 3
```
Comment (nippoo) [inline phy/datasets/mock.py]: can we try with a different n_features and n_channels to the regular ones? it's most likely to go wrong in cases with few users...

File: phy/datasets/tests/test_mock.py [hunk 2]
```diff
@@ -8,22 +8,26 @@
...
 #------------------------------------------------------------------------------
 
-def test_artificial():
-    n_spikes = 100
+def _test_artificial(n_spikes=None, n_clusters=None):
     n_samples_waveforms = 32
     n_samples_traces = 50
     n_channels = 64
-    n_clusters = 10
+    n_features = n_channels * 3
```
Comment (nippoo) [inline phy/datasets/tests/test_mock.py]: again here, let's test with 2 or 4...

File: phy/datasets/mock.py (hunk 1 above)
Comment (rossant) [inline phy/datasets/mock.py]: good idea, done

File: phy/datasets/tests/test_mock.py (hunk 2 above)
Comment (rossant) [inline phy/datasets/tests/test_mock.py]: done

**Pull request merged**